debian/paagent.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PAAGENT "1" "October 23, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
paagent \- starts ProActive nodes on a machine according to a schedule
.SH SYNOPSIS
.B paagent
.RI [ OPTIONS ]
.br
.B paagent_server
.RI [ OPTIONS ] 
.SH DESCRIPTION 
The programs are not intended to be started on their own but instead used through the 
.B
paac
and 
.B
paas
system services. The paas system service starts the
.B 
paagent_server
at boot time while paac starts
.B
paagent. 
.B
Paagent_server
will listen for method calls on DBus and start and stop nodes when it receives messages. 
.B
Paagent
reads an XML configuration file and sends start and stop commands to paagent_server. The configuration file specifies times to start, idle thresholds, resource limits, and other options.
By default 
.B
paac
uses the 
.B
config.xml
file placed in /etc/paagent to configure
.B
paagent.
Paagent
also keeps the started ProActive nodes alive by restarting them if they stop and manages the resources used (CPU, memory) by the nodes. ProActive is required to be installed for the services to run properly. A ProActive distribution can be downloaded from http://www.activeeon.com/community-downloads

.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.

.SH
paagent
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-l <filename> \-\-logging <filename>
Selects the logging configuration. The logging configuration file is installed by default in /etc/paagent/client.log4cxx for paagent and /etc/paagent/server.log4cxx for paagent_server. The logging output is in /var/log/paagent.
.TP
.B \-c <filename> \-\-configuration <filename>
Selects the XML configuration file. The configuration file is installed by default in /etc/paagent/config.xml. 
 
.SH SEE ALSO
.BR paagent_server (1)
.BR paas (1)
.BR paac (1)
.BR paagent_configuration (1)

.SH AUTHOR
paagent was written by Vasile Jureschi <vjureschi@sophia.inria.fr>.
.PP
This manual page was written by Vasile Jureschi <vjuresch@sophia.inria.fr>,
and may be used by others.


ChangeLog
2009-10-23 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com
	* removed unused files from debian directory
        menu.ex
        init.d.lsb.ex
        watch.ex
        manpage.1.ex
        manpage.xml.ex
        manpage.sgml.ex
        preinst.ex
        init.d.ex
        prerm.ex
        paagent.doc-base.EX
        paagent.default.ex
    * added manpages to the debian directory
        paas.1
        paac.1
        paagent_server.1
        paagent_configuration.1
        paagent.manpages
        paagent.1
    * fixed Section, Standard-Version, and description in debian/control
    * added entry to debian/changelog
    * fixed old address in debina/copyright
    * added ProActive requirement description in debian/README.Debian
    * added paagent.spec to the EXTRA_DIST in Makefile.am so RPM can be built
      from source distribution
    * changed version to 0.1.4
    * regenerated autotools configuration

2009-10-21 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com
   
   * fixed remove_services_fedora in Makefile.am by replacing BUILDROOT with $(DESTDIR)
   * modified the install_services_ubuntu target so it can be used with the deb target
     - added $(DESTDIR) Makefile.am for the paths in the target
   * modified the install_services_fedora to create and restore a backup of .rpmmacros
    when building the rpm
   * added deb target in Makefile.am
   * modified paas and paac to use /usr/bin instead of /usr/local/bin
   * added skeleton debian configuration in debian directory
   * regenerated autotools configuration  
	    
2009-10-16 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com
 
   * rpm building works now 
   * removed tclap include references from Makefile.am and added the tclap
     header files in the src folder
   * regenerated autotools configuration
   * added tclap references in Makefile.am

2009-10-15 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com

   * modified the paas_fedora and paac_fedora scripts to be usable on Fedora
   * changed installation directory from /etc/init.d to /etc/rc.d/init.d for
     fedora scripts (paas_fedora paac_fedora)
      - modified: Makefile.am paas_fedora paac_fedora
   * added paagent.spec file for building RPMs
   * modified the Makefile.am to take build parameters from paagent.spec
   * add make rpm target to the Makefile.am

2009-10-13 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com

   * added <jvm_memory> tag handling in NodeStarter::Initialize by adding it to
     the exec_arguments vector
   * changed this->jvm_params.empty()  to this->jvm_params.clear() in 
     Configuration::SetJVMParams, added the empty() instead of clear() by 
     mistake
   * added <use_all_cpus> handling in NodeStarter::Initialize by adding
     -native or -green to the JVM parameters
   * added <processPriority> handling in Watcher class through the renice method
     The named constructors were refactored to accept a pointer to and event and 
     an action. This was less cumbersome than passing a long list of arguments.  

2009-10-12 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com

   * added arguments parsing in CustomNodeStarter constructor by appending the 
     extra arguments to the exec_arguments vector

2009-10-12 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com
	
	* added string GetJVMParamsString() const; in Configuration class to get 
	the string of parameters
   
2009-10-12 Vasile Jureschi  vjuresch@sohia.inria.fr, vasile.jureschi@gmail.com
	* added namespace "paagent" to prevent name clashes (e.g. Event class)
	 
	* fixed the cpulimiting, as it was not implemented for every action 
		- modified Runner class 
	
	* modified the starter classes (Runner, Controller, NodeStarter 
	  and derived classes) to use a simpler way of passing arguments. 
	  In the paagent_server the configuration options are now
	  passed around using a Configuration singleton class. 
	
	* added handling for the <jvm_args> option in the configuration xml
	    - modified the Configuration class and NodeStarter class
   

paagent.cpp
/*
 * ################################################################
 *
 * ProActive: The ProActive Linux Agent
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 * USA
 *
 * ################################################################
 */

#include "Runner.h"
#include <iostream>
#include <string>
#include <tclap/CmdLine.h>
#include <log4cxx/logger.h>
#include <log4cxx/propertyconfigurator.h>
#include <log4cxx/helpers/exception.h>
#include <log4cxx/file.h>

using namespace log4cxx;
using namespace log4cxx::helpers;
using namespace std;

/**
 * Entry point for the ProActive Linux Agent command line client.
 * The main function parses the command line arguments using the
 * tclap library and search for a configuration file for the agent
 * and a configuration file for logging.
 *
 * USAGE:
 *
 * ./paagent  [-l <string>] -c <string> [--] [--version] [-h]
 *
 *
 * Where:
 *
 * -l <string>,  --logging <string>
 * Log4CXX configuration file
 *
 * -c <string>,  --configuration <string>
 * (required)  Configuration file
 *
 * --,  --ignore_rest
 * Ignores the rest of the labeled arguments following this flag.
 *
 * --version
 * Displays version information and exits.
 *
 * -h,  --help
 * Displays usage information and exits.
 *
 *
 * Starts the ProActive Linux Agent. The Agent will read an XML
 * configuration file and start and stop nodes according to an event
 * schedule or to idle thresholds.
 *
 * @param arc the name of the program
 * @param argv the arguments passed to the program, configuration and logging file
 * @see paagent_server.cpp
 * @author vjuresch
 * @since 0.1
 */
int main(int argc, char** argv) {

	try {
		//define description, delimiter, and version
		TCLAP::CmdLine cmd("Starts the ProActive Linux Agent. The Agent will "
			"read an XML configuration file "
			"and start and stop nodes according to an event "
			"schedule or to idle thresholds.", ' ', "0.1");

		//define configuration argument
		TCLAP::ValueArg<std::string> xmlFileArg("c", "configuration",
				"Configuration file", true, "", "string");
		//define logging argument
		TCLAP::ValueArg<std::string> log4cxxFileArg("l", "logging",
				"Log4CXX configuration file", false, "", "string");

		//add to command line object
		cmd.add(xmlFileArg);
		cmd.add(log4cxxFileArg);
		// Parse the argv array.
		cmd.parse(argc, argv);

		//get the values
		string xml_file = xmlFileArg.getValue();
		string log4cxx_file = log4cxxFileArg.getValue();

		//arguments logic
		LoggerPtr logger = log4cxx::Logger::getRootLogger();
		//if log4cxx_file is set try to load, else do a basic config
		if (log4cxxFileArg.isSet()) {
			PropertyConfigurator::configure(log4cxx_file);
			LOG4CXX_INFO(logger, "Logging initialized from file: " << log4cxx_file);

		} else {
			BasicConfigurator::configure();
			//setting level to info since not specifying a logging file
			//means the user is only interested only in basic information
			logger->setLevel(log4cxx::Level::getInfo());
			LOG4CXX_INFO(logger, "Logging initialized with basic configuration.");
		}

		//start agent client
		Runner runner(xml_file);
		LOG4CXX_DEBUG(logger, "Starting agent main loop...");
		runner.start();
		runner.join();

	} catch (TCLAP::ArgException &e) {
		std::cerr << "error: " << e.error() << " for arg " << e.argId()
				<< std::endl;
	}
}

